### PR TITLE
Add iterator interface to detected_object_type

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
Add an iterator/range interface (i.e. begin and end methods) to `detected_object_type`. Besides convenience, this provides a much more efficient mechanism to visit all classifications, as previously the only way to do so was to iterate over all class names (which was expensive due to creating a temporary vector and sorting by score) and then look up (O(log N)) the score for each class name.

### CODE REVIEW

**Comment by mwoehlke-kitware** [comment]
Caveat: the iterators are `pair<string const*, double>`, not `pair<string, double>` (note the indirection of the `string`). This is because I am returning the iterators to the internal map directly.

This is mildly inconvenient, but I don't *think* exposing the indirect mapping is a concern since the strings are immutable. (Also, it potentially saves making a copy of the string.) However, if this is considered a problem, it should be possible to make an iterator adapter to hide this detail.

File: vital/types/detected_object_type.h [hunk 1]
```diff
@@ -74,6 +74,9 @@ class VITAL_EXPORT detected_object_type
 public:
   static const double INVALID_SCORE;
 
+  using class_map_t = std::map< std::string const*, double >;
```
Comment (linus-sherrill) [inline vital/types/detected_object_type.h]: Could this be a std::string const& rather than a pointer?

Comment (mwoehlke-kitware) [inline vital/types/detected_object_type.h]: > he just changed it to a typedef (using) so it was clearer.

...or, so I could write the iterator type without it being obscenely long :wink:. But yes, the implementation uses pointers to strings (the instances are stored in the master types table, I believe), so unfortunately this implementation detail is exposed. See my previous comment; if this is an issue, I believe an iterator adapter can be created to hide it...

**Comment by mleotta** [comment]
I see no issue with this.

**Pull request merged**